### DISCUSSION TITLE
iOS Room Assistant Companion App

### DISCUSSION THREAD

**Discussion by julianrinaldi** [mKeRix/room-assistant#352] [Q&A]
I've installed the Room Assistant companion app, but Room Assistant isn't finding my iPhone 11 Pro with iOS 14.2. I originally put in the UUID in all capital letters because that's what the app showed. When that didn't work, I remembered it was supposed to be in lowercase, so I changed it. That also didn't work. Then I realized I shouldn't have put the dashes from the UUID in. Still nothing. Other ibeacons I have work fine. Obviously I have bluetooth enabled for the companion app. Is there a trick to getting it to work?

**Comment by mKeRix** [comment]
Hey there, actually for the companion app the ID needs to be exactly as stated in the app (all uppercase with dashes) to work. I know this is a bit confusing and already noted to make it case insensitive.

Do you see any messages like "attempt app discovery" in your logs? Could you maybe check if there is something notable while running room-assistant in verbose mode? If you also have another iOS device like an iPad you can check from that if the phone broadcasts the right data as well, let me know if that is something you want to try.

**Reply by julianrinaldi** [reply]
Let me try the ID again. If it doesn't work I can try what you suggest with the iPad.

**Reply by julianrinaldi** [reply]
No luck. Which logs should I be looking at?

**Reply by mKeRix** [reply]
Just so that we're on the same page - are you running the beta of the room-assistant software (not the app!) as well? See also what I commented down below.

If yes: the logs of the room-assistant instance would be interesting. From your iPad you can also do a sanity check by downloading the nRF Connect app, scanning and connecting to your iPhone. When you swipe over in the resulting device details menu you should see a bunch of services and characteristics. The app should have created a service (ID starting with `5403`) and given that service a characteristic (starting with `21c4`). Can you see those two and read the value from the characteristic (click the little down arrow on it)?

**Reply by julianrinaldi** [reply]
I wasn't running the beta. I'm working on installing it now.

The value I'm getting is:
`0x36-38-44-41-34-30-41-31-2D-38-46-39-44-2D-34-39-31-42-2D-39-36-38-45-2D-37-45-30-39-39-33-33-39-31-46-33-30`

**Reply by julianrinaldi** [reply]
MY issue was obviously not installing Room Assistant beta, because I'm seeing it now. I asked in another location, but is there an easy way to install the beta to all my nodes through ansible?

**Reply by mKeRix** [reply]
You can specify the specific version or `beta` in the  version variable to use the beta with ansible.

**Reply by julianrinaldi** [reply]
I changed "room_assistant_version" in ansible-playbooks/roles/room-assistant/defaults/main.yml to "beta" instead of "latest". Is this the correct way to do it?

**Reply by mKeRix** [reply]
Exactly, afterwards run the ansible playbook and it will update all instances to the beta. :)

**Comment by meme1337** [comment]
I want to add my experience to this issue, as I'm facing the same behavior.

I've installed the companion app on my iPhone Xr (iOS 14.2). Bluetooth permission is granted.
Room-assistant is installed on a raspberry pi zero w, following the guide on the site.

Room-assistant starts correctly, and discovers a lot of BLE devices, but not the iPhone that is whitelisted with the code provided in the companion app.
If I check mosquitto logs I cannot see the topic for the device_tracker, but room-assistant is working because I can see the 2 topics for the cluster leader and cluster size.

Is there something that I'm missing?

**Reply by mKeRix** [reply]
Do you see any logs stating "attempting companion app discovery"? If not: are you running the beta of room-assistant (not just the app, actual room-assistant) as well? It will not be installed by default, you need to specifically install it with e.g. `sudo npm i --global --unsafe-perm room-assistant@beta` or just choose the appropriate version of the Docker image or Home Assistant add-on.

**Reply by meme1337** [reply]
Thank you! That's definitely one issue. I will install it right now! 

And thank you for the immediate reply! I didn't expect it on a Sunday! You're awesome 👍 

p.s. can I just run the command above over the normal app? Or better to uninstall first with npm uninstall?

**Reply by mKeRix** [reply]
You can just run it over the normal app and restart it, no uninstall necessary. npm will manage everything for you. :)

**Reply by meme1337** [reply]
Thank you! Just tried, got these warnings at the end of the install

`npm WARN @nestjs/mapped-types@0.1.1 requires a peer of class-transformer@^0.3.0 but none is installed. You must install peer dependencies yourself.`
`npm WARN @nestjs/mapped-types@0.1.1 requires a peer of class-validator@^0.11.1 || ^0.12.0 but none is installed. You must install peer dependencies yourself.`
`npm WARN ws@7.4.0 requires a peer of bufferutil@^4.0.1 but none is installed. You must install peer dependencies yourself.`
`npm WARN ws@7.4.0 requires a peer of utf-8-validate@^5.0.2 but none is installed. You must install peer dependencies yourself.`

But I don't see the message "attempting companion app discovery" in the log. I'll try also in verbose mode.

**Reply by mKeRix** [reply]
The warnings don't impact room-assistant in any way, I just can't get rid of them without introducing unnecessary other dependencies atm. 

You can also try to open the app on your phone and then restart room-assistant once more while the app is open. This sometimes can help kickstart the iPhone broadcasting, at least from what I noticed in my tests. During normal operation the app does not need to be open though.

**Reply by meme1337** [reply]
I've tried this, but no results.
I'll try rebooting both the phone and the pi, I will update with pasted logs too after this attempt if it bears no result.

**Reply by meme1337** [reply]
my config
```
global:
  instanceName: <instanceName>
  integrations:
    - homeAssistant
    - bluetoothLowEnergy
homeAssistant:
  mqttUrl: 'mqtt://<mqtt-server>:1883'
  mqttOptions:
    username: <user>
    password: <password>
bluetoothLowEnergy:
  whitelist:
    - <id-from-companion-app>
```
```
11/22/2020, 5:03:44 PM - info - IntegrationsModule: Loading integrations: home-assistant, bluetooth-low-energy
11/22/2020, 5:03:50 PM - info - NestFactory: Starting Nest application...
11/22/2020, 5:03:50 PM - info - InstanceLoader: AppModule dependencies initialized
11/22/2020, 5:03:50 PM - info - InstanceLoader: ConfigModule dependencies initialized
11/22/2020, 5:03:50 PM - info - InstanceLoader: NestEmitterModule dependencies initialized
11/22/2020, 5:03:50 PM - info - InstanceLoader: IntegrationsModule dependencies initialized
11/22/2020, 5:03:51 PM - info - InstanceLoader: HttpModule dependencies initialized
11/22/2020, 5:03:51 PM - info - InstanceLoader: DiscoveryModule dependencies initialized
11/22/2020, 5:03:51 PM - info - InstanceLoader: ClusterModule dependencies initialized
11/22/2020, 5:03:51 PM - info - InstanceLoader: TerminusModule dependencies initialized
11/22/2020, 5:03:51 PM - info - InstanceLoader: BluetoothModule dependencies initialized
11/22/2020, 5:03:51 PM - info - InstanceLoader: ScheduleModule dependencies initialized
11/22/2020, 5:03:51 PM - info - InstanceLoader: EntitiesModule dependencies initialized
11/22/2020, 5:03:51 PM - info - InstanceLoader: HomeAssistantModule dependencies initialized
11/22/2020, 5:03:51 PM - info - InstanceLoader: BluetoothLowEnergyModule dependencies initialized
11/22/2020, 5:03:51 PM - info - InstanceLoader: StatusModule dependencies initialized
11/22/2020, 5:03:54 PM - info - RoutesResolver: EntitiesController {/entities}:
11/22/2020, 5:03:54 PM - info - RouterExplorer: Mapped {/entities, GET} route
11/22/2020, 5:03:54 PM - info - RoutesResolver: StatusController {/status}:
11/22/2020, 5:03:54 PM - info - RouterExplorer: Mapped {/status, GET} route
11/22/2020, 5:03:56 PM - info - HomeAssistantService: Successfully connected to MQTT broker at mqtt://192.168.1.59:1883
11/22/2020, 5:03:56 PM - info - ConfigService: Loading configuration from /usr/local/lib/node_modules/room-assistant/dist/config/definitions/default.js, config/local.yml
11/22/2020, 5:03:57 PM - info - ClusterService: Starting mDNS advertisements and discovery
11/22/2020, 5:03:57 PM - info - NestApplication: Nest application successfully started
11/22/2020, 5:03:57 PM - info - BluetoothLowEnergyService: Discovered new BLE peripheral 2be11f59870d with ID 2be11f59870d and RSSI -33
11/22/2020, 5:03:57 PM - info - BluetoothLowEnergyService: Discovered new BLE peripheral 7f2a55d0108c with ID 7f2a55d0108c and RSSI -48
11/22/2020, 5:03:57 PM - info - BluetoothLowEnergyService: Discovered new BLE peripheral f83f513aa8bb with ID f83f513aa8bb and RSSI -65
11/22/2020, 5:03:57 PM - info - BluetoothLowEnergyService: Discovered new BLE peripheral lgtv with ID 6b9bf51f7817 and RSSI -65
11/22/2020, 5:03:57 PM - info - BluetoothLowEnergyService: Discovered new BLE peripheral 515754b5fdf7 with ID 515754b5fdf7 and RSSI -70
11/22/2020, 5:03:57 PM - info - BluetoothLowEnergyService: Discovered new BLE peripheral 496020a5f2e1 with ID 496020a5f2e1 and RSSI -44
11/22/2020, 5:03:58 PM - info - BluetoothLowEnergyService: Discovered new BLE peripheral SHIELD with ID 71c281a92d0b and RSSI -67
11/22/2020, 5:03:58 PM - info - BluetoothLowEnergyService: Discovered new BLE peripheral 710d5dcc4cfb with ID 710d5dcc4cfb and RSSI -64
11/22/2020, 5:03:58 PM - info - BluetoothLowEnergyService: Discovered new BLE peripheral 00c3f4e2c2df with ID 00c3f4e2c2df and RSSI -92
11/22/2020, 5:03:59 PM - info - BluetoothLowEnergyService: Discovered new BLE peripheral 2811a5ba07f2 with ID 2811a5ba07f2 and RSSI -77
11/22/2020, 5:04:00 PM - info - BluetoothLowEnergyService: Discovered new BLE peripheral 6e116e659aeb with ID 6e116e659aeb and RSSI -89
11/22/2020, 5:04:02 PM - info - BluetoothLowEnergyService: Discovered new BLE peripheral fc039f594e55 with ID fc039f594e55 and RSSI -97
11/22/2020, 5:04:02 PM - info - BluetoothLowEnergyService: Discovered new BLE peripheral 691b02b89ae4 with ID 691b02b89ae4 and RSSI -73
```

**Reply by mKeRix** [reply]
Hm, the logs look ok to me. Could you try if [this fix](https://github.com/mKeRix/room-assistant/issues/309#issuecomment-726015361) from the other thread changes anything for you? Just turn it down to `2` or remove the line completely.

**Reply by meme1337** [reply]
I've tried with 2 and removing completely, but no luck. It's not showing anything related to the companion app :(
Let me know if I can do other tests, I'm willing and available to be a guinea pig :)

Thank you!

**Reply by julianrinaldi** [reply]
> You can just run it over the normal app and restart it, no uninstall necessary. npm will manage everything for you. :)

Is there an easy way to do this with ansible? I have it set up through that. I saw a file with "room_assistant_version" in it. Can I just change that to "beta" from "latest"?

**Reply by mikepflu** [reply]
I think I'm having the same issue as meme1337.  I installed the companion app on my iPhone, updated room-assistant to the beta version.  I took the device id from the room-assistant companion app and put it in the whitelist in my bluetoothLowEnergy section.  I even removed the check in bluetooth-low-energy.service.js for manufacturer data length, and I'm not seeing my iPhone show up in the room-assistant logs, running verbose.

**Reply by mKeRix** [reply]
The confusing thing is: with that condition removed it should try to connect to any advertising device manufactured by Apple, so you should at least see the attempts being logged. Do either of you own an iPad that you could use for checking if the iPhone appears as advertising device when scanning in the nRF Connect app? Also, did you ever connect something via Bluetooth to your iPhones (e.g. headphones, AirPods, Apple Watch...)?

**Reply by meme1337** [reply]
I don’t have an iPad to check, but the iPhone was connected to the Apple Watch during all my attempts.

**Reply by mikepflu** [reply]
I have an ipad and I installed the nRF Connect app, but I've never used it before.  I do not see anything that looks like my iphone when scanning in the app and I have paired my iPhone with multiple devices, headphones, car, etc...  When scanning with nRF Connect, what would I look for?  For example, would I look for my bluetooth MAC, the address from the companion app, the name of my iPhone?

**Reply by mikepflu** [reply]
Using the scanner, I see devices like my wife's apple watch, the macbook pro that I'm using right now, my wife's phone and some hue light bulbs that I have, as well as other devices that say N/A, but I don't see anything like my phone.  Maybe it's not broadcasting?

**Reply by meme1337** [reply]
Is there a way for me to check that the iPhone is broadcasting without an iPad or another Apple device?

**Reply by meme1337** [reply]
as OP solved his issue, I will create a new discussion related to my use case.

**Reply by mKeRix** [reply]
@mikepflu Hm, that's unfortunate. For devices that are signed into the same iCloud account the device should just show up as whatever the device name is (e.g. "Mike's iPhone") afaik, at least that's what it was like for me. I did have some patches where my iPhone was undiscoverable by my iPad as well during development, however that was usually solved by toggling Bluetooth off/on in the settings or a reboot. I tried to google around a bit, but I also could not find any other reports of iOS devices not broadcasting BLE at all. :/

**Reply by meme1337** [reply]
@mKeRix Hi and sorry to ping you directly. I think this saturday I will be able to get my paws on some iDevices. What should I do to check if my Xr is broadcasting BLE?

**Reply by mikepflu** [reply]
@mKeRix Sorry I'm just getting back to looking at this.  I had been doing some things with bluetooth on my first iPhone and found a paired device with what looked like just a MAC.  It's possible that the pairing was my Raspberry Pi, so I removed it.  Since then, when I run nRF Connect on a spare iPhone (broken screen), I now see a device in the list called "Mike".  I checked the name on my phone, and it matches, so I'm guessing that that's the device, yay!  Maybe it was a reboot, or me disabling / enabling bluetooth, or just removing that pairing, I don't know, but I think I'm good on my device broadcasting BLE.  So, now that it appears to be broadcasting, what do I look for next?

W/hen I look at MQTT, I don't see anything in there that looks like my device.  Also, running the process directly (not through systemctl) with the command "/usr/bin/room-assistant -v -c /home/pi/room-assistant/config", I don't see anything in my shell session that looks like my phone.

Here's the contents of my config file:
```
$ cat config/local.yml
global:
  instanceName: Living Room
  integrations:
    - homeAssistant
    - bluetoothLowEnergy
homeAssistant:
  mqttUrl: mqtt://192.168.50.20:1883
bluetoothLowEnergy:
  hciDeviceid: 0
  whitelist:
    - 0D9B8B29-2497-4BA8-9E07-1DE56900706E
```

I'm sitting about 12 feet away from the Raspberry Pi with my iPhone right next to me, so I would expect that it should see it.  I feel like I'm missing something.

**Reply by mikepflu** [reply]
So, I just tried something on my iPhone to be sure that I'm seeing the proper device broadcasting.  I disabled bluetooth, changed the name on my phone to "Mike-iPhone" and enabled bluetooth.  In the scanner, I'm still seeing a device "Mike", and still nothing in room-assistant.  I don't know what I should be looking for in the scanner to know that it's actually broadcasting.

**Reply by mikepflu** [reply]
OK, sorry to keep narrating this as it's happening, but I feel like I'm getting closer.  I did a few more scans and I finally found the newly renamed device "Mike-iPhone".  Reading over older posts, I realized that I need to "Connect" to the device, and when I look under the services tab at the top, I see Advertised Services and Attribute Table.  Under Advertised Services, it says "This device does not advertise any service."  Under Attribute Table, there are a lot of entries.  To me, it seems like there is no advertised service, and I don't see anything that looks like the string on my iphone, or anything starting with 5403 or 21c4.

I feel like I'm getting closer, but I'm not quite there yet.

**Reply by mikepflu** [reply]
So, I guess being patient is what matters.  It seems like I'm now seeing the service, so here's a screenshot from the scanning app, and a screenshot from the room assistant on my iPhone.

![IMG_55981B50DA56-1](https://user-images.githubusercontent.com/36419383/100398875-8268f200-301e-11eb-994a-4e7d05446fc2.jpeg)
![IMG_502DD0C4950C-1](https://user-images.githubusercontent.com/36419383/100398898-a0ceed80-301e-11eb-99a7-3236aea761cb.jpeg)

**Reply by mikepflu** [reply]
And to round out my night, on a whim, I tried rebooting my raspberry pi, and I think it's working now!!!

```
Nov 27 00:55:08 monitor-first-floor room-assistant[354]: 11/27/2020, 00:55:08 - info - BluetoothLowEnergyService: Discovered companion app with ID 0D9B8B29-2497-4BA8-9E07-1DE56900706E for tag 54b8cf2917ea
Nov 27 00:55:08 monitor-first-floor room-assistant[354]: 11/27/2020, 00:55:08 - info - BluetoothLowEnergyService: Discovered new BLE peripheral 0D9B8B29-2497-4BA8-9E07-1DE56900706E with ID 0D9B8B29-2497-4BA8-9E07-1DE56900706E and RSSI -57
Nov 27 00:55:08 monitor-first-floor room-assistant[354]: 11/27/2020, 00:55:08 - info - HomeAssistantService: Device tracker requires manual setup in Home Assistant with topic: room-assistant/device_tracker/ble-0d9b8b29-2497-4ba8-9e07-1de56900706e-tracker/state
```

And now I seem to have a device and an entity in HA under MQTT, so I'm going to say that this is a success, at least for the one room.  Tomorrow I'll be setting up my Pi Zero with Ethernet and room-assistant to see if I can get multi room presence detection.

This is awesome.
Any idea when we might see a non-beta version of the companion app?

**Reply by meme1337** [reply]
@mikepflu I'll try reproducing some of your steps, but I never paired the phone with the pi, I don't think that should be needed as BLE is different. But I'll tinker with it during the weekend

**Reply by meme1337** [reply]
@mKeRix Absolutely no success. I couldn't try with the app to see if I'm transmitting, but I don't see why it shouldn't be the case.
I've reinstalled the beta of the app on the pi zero w and rebooted the phone many times.
Even uninstalled and reinstalled the companion app on the phone.

Is there any log that I can provide, or some way to see what is going wrong?

**Comment by mikepflu** [comment]
I'm not trying to hijack this discussion, but what is the relationship between room-assistant and the HA companion app?  What does it have to do with the beta version of room-assistant?  I'm trying to get all of this working together and stumbled upon this conversation.  Is there an explanation somewhere of what the benefits might be and how to set it up?  Thanks!

**Comment by julianrinaldi** [comment]
Mike, we're talking about the new beta Room Assistant companion app, not the Home Assistant companion app.

**Reply by mikepflu** [reply]
Where can I find more information about the Room Assistant companion app?  Is it something that I would run on an iPhone?

**Reply by julianrinaldi** [reply]
https://testflight.apple.com/join/OpCkeHf8

Its only purpose is to run on an iPhone. All it does is give the iPhone the ability to work with RA as a Bluetooth Low Energy beacon.

**Reply by mikepflu** [reply]
Are there any docs on setting it up.  I wouldn't mind beta-testing, but I probably need a little more info on what needs to be configured.  I already have TestFlight from some other testing I did previously.

**Reply by julianrinaldi** [reply]
You need to install the beta version of Room Assistant on the raspberry pis you're using with RA. Stop the RA service, then upgrade: `sudo npm i --global --unsafe-perm room-assistant@beta` Then restart RA. Then install the app through test flight. It will give you a MAC address. Put that address in your Room Assistant configuration exactly how it's written, with uppercase letters and dashes.

**Comment by meme1337** [comment]
No success. I tried so many times, reinstalling the beta over, reinstalling the companion app, making sure that BT is enabled and permitted to the app.

Nothing works. I don't know what to do to investigate to be honest :(

**Reply by mKeRix** [reply]
Unfortunately I'm really not sure what the issue is as well. :/ Your logs don't suggest anything being wrong. Which iPhone and iOS version are you trying this on? 

You can see the currently installed version of room-assistant by running `room-assistant -h`, which will print the version number in the heading. This should be 2.13.0-beta.3.

If you run `sudo hcitool cc <phone-mac> && sudo hcitool rssi <phone-mac>` on your Pi, what does this return? Replace `<phone-mac>` with the Bluetooth MAC address of your iPhone, which you can find under Settings > General > Info > Bluetooth. This will test if the iPhone responds to Bluetooth Classic inquiries - which is different from the mechanism used by the companion app, but still might give us some insight.

**Reply by meme1337** [reply]
> Unfortunately I'm really not sure what the issue is as well. :/ Your logs don't suggest anything being wrong. Which iPhone and iOS version are you trying this on?
> 
> You can see the currently installed version of room-assistant by running `room-assistant -h`, which will print the version number in the heading. This should be 2.13.0-beta.3.
> 
> If you run `sudo hcitool cc <phone-mac> && sudo hcitool rssi <phone-mac>` on your Pi, what does this return? Replace `<phone-mac>` with the Bluetooth MAC address of your iPhone, which you can find under Settings > General > Info > Bluetooth. This will test if the iPhone responds to Bluetooth Classic inquiries - which is different from the mechanism used by the companion app, but still might give us some insight.

Thank you for your reply!

I am using an iPhone Xr with latest iOS 14.2.

These are the info from the commands above:

`room-assistant -h` returns 2.12.0. 
This is very weird, as I reinstalled it again today with the command for beta, as follows:

`sudo npm i --global --unsafe-perm room-assistant@beta`

I’m thinking that there is something “dirty” from my first attempt a couple weeks ago , as my first one was the normal version, not the beta. 
That would explain also why ble is not working!

Do you know how could I uninstall it completely in order to reinstall it?

If I try to run `sudo hcitool cc <phone:mac>` the phone is not connected (I was on the bt pairing screen on the mobile device), and the following command `sudo hcitool rssi <phone:mac>` states so as well.

Today, during other attempts, I used bluetoothctl, and was able to pair the device successfully. It has now been unpaired too.

If I run `sudo hcitool scan on` I can see the phone amongst the other bt devices in range.

**Reply by mKeRix** [reply]
`sudo npm uninstall -g room-assistant` should delete room-assistant and all its dependencies from your machine. After running that I would double-check that the `room-assistant` command does not work anymore, just to be safe that there is no other weird stuff going on with npm. After that you should be able to re-install with the command you posted.

**Reply by meme1337** [reply]
> `sudo npm uninstall -g room-assistant` should delete room-assistant and all its dependencies from your machine. After running that I would double-check that the `room-assistant` command does not work anymore, just to be safe that there is no other weird stuff going on with npm. After that you should be able to re-install with the command you posted.

Run the command, got a message for 425 package removed, but if I run room-assistant it still starts the program...
This is incredibly weird.

**Reply by meme1337** [reply]
Is there a way to manually remove everything? It seems there is still something wrong on my pi.

**Reply by mKeRix** [reply]
`which room-assistant` should show you the file that it uses to start room-assistant. If you copy the path it gives you and insert it into `ls -la <insert-path>` it should show you which path that links to (if it is a link). You can delete the room-assistant folder and all its contents from the path of the second command, and the file from the first command. That should wipe it entirely.

**Reply by meme1337** [reply]
Heheheh I was doing almost the same. I run it and from ps aux searched the path. 
I found only a script with +x in /usr/local/bin and I deleted it. I’m now installing the beta again. 
Crossing fingers!

**Reply by meme1337** [reply]
It worked!!

Thank you again @mKeRix !

I have some questions about room-assistant, just to be sure I understand how it works with multiple instances:

- if two or more instances will see the beacon in range, will the one with higher signal reports presence? Or how does it work?
- As I don’t have a big house, can I theoretically just use one or two and use the signal strength in a template sensor to guess which room I’m in? 

Thanks again for your time and support!

**Reply by mKeRix** [reply]
When multiple instances are in play room-assistant will set the room presence sensor to the instance that is:

- has the lowest distance value
- is not considered out of range (can be controlled with `maxDistance`)
- has a recent measurement (can be controlled with `timeout` - I recommend to raise this setting to e.g. 60 for the companion app)

For the template sensor, in theory you could, but it's not straightforward. You will have to work with the REST sensors of Home Assistant (or build some other way to get the data from the HTTP API) and then do your calculation based on the distances you grab there. Somebody posted about how they did triangulation with room-assistant [here](https://community.home-assistant.io/t/internal-position-tracking-with-co-ordinates-using-an-iphone-and-room-assistant/254399), that is with Node-RED though.

In the future I want to remove the requirement of having an instance per room by doing something similar to FIND3, but that is still very much in the concept phase.
